### PR TITLE
ci: add cache for ClamAV and use dedicated script for package installation

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

This PR does two things:

1. Add build caching for ClamAV's bytecode
2. Use a separate script for installing packages

The separate script provides a better overview of the installation process. While I was at it, I removed non-essential packages from the packages list. Therefore, this PR is a breaking change.

Moreover, users that want to build DMS themselves are now required to use a recent version of Docker because we need BuildKit (one needs Docker version `>= 20.10.14`).

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by georglauterbach** [comment]
I noticed that this PR supersedes #2747. #2747 may be closed.

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:3]: I'm not too good with memory, but I'm probably the odd one out here with active maintainers currently. Is there any value including a brief comment that describes what is accomplished? (Optional)

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:9]: Is this the only place that `DEBIAN_FRONTEND` and `LOG_LEVEL=trace` matters?

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:33]:
```suggestion
  _log 'warn' 'Applying workaround for Postfix bug (see https://github.com/docker-mailserver/docker-mailserver/issues/2023#issuecomment-855326403)'
```

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:37]: Just as a note, I think `docker build` offers a `--hostname` arg now which might allow to remove the workaround. Don't worry about that for this PR though.

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:55]: This is new and looks like it could be deferred to a separate PR (_better for `git blame` / context_).

My preference (_unless you know of an issue otherwise_) would be to have `rspamd` managed via it's own stage, like `clamav` and `fail2ban` should be. For now though, I think this is fine, but should be introduced via a separate PR.

File: target/scripts/build/packages.sh
Comment (polarathene) [inline target/scripts/build/packages.sh:83]: Likewise, it would be good to separate packages by features. It's not always clear why we have some packages (_and with time gets harder to track via `git blame`_).

- I know that `uuid` was added for the current file locking scripts.
- `clamav` + `clamav-daemon` could be by itself or for the time being mixed with `amavisd-new` + `spamassassin`. I think `p7zip-full` is for `amavisd-new` support.
- `libldap-common`, `postfix-ldap`, `libsasl2-modules`, `sasl2-bin` for LDAP (_and Cyrus based SASL? One might not be_ :man_shrugging: )
- `redis-server` stood out. This DB package is for `rspamd` I assume? (_highlights why I'd like the separation / scoping to features/dependencies_)
- You've dropped quite a few packages...why? (**EDIT:** [Mentioned in this commit](https://github.com/docker-mailserver/docker-mailserver/pull/2752/commits/f9961baf2861ad53e5a421b1ad0e0621c27a79ec)) These provide support to Amavis AFAIK, it lists all the supported codecs found during the startup logs, ClamAV maybe also uses it to check for viruses? Is this something you can confidently drop? (_with proper stage separation, we can halve the weight of the image if we publish an alternative antivirus tag variant, or the other way around with a lightweight `core` base_)

Some of the packages dropped aren't related to the archive support that the commit messages justifies the removal with. **I am not comfortable this being part of the PR.**

Please maintain the existing package list as-is, and defer the `rspamd` ones to a separate PR for introducing `rspamd` support. A separate PR may follow this one to introduce the breaking change of removing packages, and that is all that PR should cover, preferably the scoping/grouping would already be laid out here.

File: target/scripts/helpers/log.sh
Comment (polarathene) [inline target/scripts/helpers/log.sh:58]: No need to squeeze this into one line. I don't mind it when it's declaration only, but not a fan with assignment involved:

```suggestion
  local LEVEL_AS_INT
  local MESSAGE="${RESET}["
```

File: target/scripts/helpers/log.sh
Comment (polarathene) [inline target/scripts/helpers/log.sh:121]: I tried to follow what `LOG_LEVEL:1:-1` is intended for since you compare to strings... I looked at the original PR [commit that introduced this](https://github.com/docker-mailserver/docker-mailserver/pull/2506/commits/f0668344379b742839ccaf64fa4f1834815164fc), but didn't see anything to explain that either.

I assume it's for triggering the `*` case?

File: target/scripts/helpers/log.sh
Comment (polarathene) [inline target/scripts/helpers/log.sh:45]: I saw your [associated commit message](https://github.com/docker-mailserver/docker-mailserver/pull/2752/commits/d5e33f5f9ac36325a2f43d66f635ab8da0486adf) for this, but I'm not sure what you're referencing?

We generally keep to that style now don't we? Has there been complaints somewhere? I think the only case that doesn't happen is with `.bats`, or in my case if I forget to style it that way and it slips through review (_in languages I'm used to, it's common to take this approach with `{` on the same line_).

Personally, unless there's a good reason to, we should already be pretty consistent with the function style used, I don't think we need to slowly lose that consistency (_or do a big PR to change it_). If another maintainer wants that, I'm fine with it, but it is a gripe I have when trawling history via `git blame` for more information on something to wade through these kind of broad sweeping adjustments.

**Review by polarathene — CHANGES_REQUESTED** [review]

File: .github/workflows/default_on_push.yml
Comment (polarathene) [inline .github/workflows/default_on_push.yml:50]: In reference to [associated commit message](https://github.com/docker-mailserver/docker-mailserver/pull/2752/commits/05c533e8dd0a065367c1b995522be2d4fd7a128a):

I cannot track down the PR (perhaps it was a direct commit) that introduced this for more information.

Only where it was introduced before these workflow files with just [`Makefile` usage](https://github.com/docker-mailserver/docker-mailserver/pull/1256#issuecomment-529406724). It looks like whomever added this to the workflows misunderstood it?

I don't see why `VCS_VERSION` is assigned `github.ref`, but maybe my memory is bad and that will use a tag when it's a tag event. Otherwise AFAIK it would just be the branch ref it's running from.

It's intention was to just bake in the info into image metadata so it could be queried and look like:

```
org.label-schema.vcs-ref=7580aec
org.label-schema.version=v6.1.0-111-g7580aec
```

---

`VCS_REF` probably referred to [`git ref`](https://git-scm.com/book/en/v2/Git-Internals-Git-References). That's the commit SHA for the inverse of `github.ref` that `VCS_VER` has been using here (_while the `Makefile` instead uses [`git describe`](https://git-scm.com/docs/git-describe), so isn't consistent_).

**Comment by polarathene** [comment]
Sorry for lack of overview / summary, browser was acting up and I thought I almost lost the review before submitting so rushed it :sweat_smile:

I don't have enough time to provide one right now :weary: I also may not be as available until later in the week/weekend.

> 2\. Use a separate script for installing packages

I am not sure if this is something I've seen in `Dockerfile` following best practices. I am not sure if that causes different behaviour for `RUN` caching? The directive itself won't change, but I think when the file itself does it's ok, so it should be fine?

File: Dockerfile
Comment (casperklein) [reply Dockerfile:25]: I was about to suggest the same, use `ARG` here 👍

File: Dockerfile
Comment (casperklein) [reply Dockerfile:42]: I think, copying the database from the official image is a good idea, instead of running `freshclam` on our own. Running freshclam is rate limited. I didn't check it, if that was already the case in the past, but theoretically we can end with an image without a current DB due to rate limits by clamav. At least while testing/building locally, I run very quickly in that rate limit.

File: target/scripts/build/packages.sh
Comment (casperklein) [reply target/scripts/build/packages.sh:55]: `rspamd` is not scope of that PR and should be done separately.

**Comment by casperklein** [comment]
It was said somewhere else already and I second this again: To make reviews easier, keep them as small as possible and avoid changes not directly needed for the goal a PR achieves (such as beautifying things or introducing other little additions like rspamd).
Like the unix philosophy: do one thing well 👍

**Comment by georglauterbach** [comment]
I see, my PR became a bit bit and not UNIX-oid.. I will provide smaller ones that incorporate the review feedback - thank's @polarathene :)

File: Dockerfile
Comment (georglauterbach) [reply Dockerfile:25]: 👍🏼

File: Dockerfile
Comment (georglauterbach) [reply Dockerfile:42]: Nice observation. I will see what I can do.

File: Dockerfile
Comment (georglauterbach) [reply Dockerfile:280]: Just style, can be removed. I will not move them in upcoming PRs.

File: Makefile
Comment (georglauterbach) [reply Makefile:16]: 👍🏼

File: VERSION
Comment (georglauterbach) [reply VERSION:1]: This is indeed an accident and should not have happened :D

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:3]: I can provide one :)

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:9]: Yes, so we can probably remove them entirely from the Dockerfile of we use a dedicated script.

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:33]: 👍🏼

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:55]: Agreed!

File: target/scripts/build/packages.sh
Comment (georglauterbach) [reply target/scripts/build/packages.sh:83]: I see. I will not touch the packages :)

PS: You're right: `redis-server` is for `rspamd`

File: target/scripts/helpers/log.sh
Comment (georglauterbach) [reply target/scripts/helpers/log.sh:45]: Agreed. I always thought people hate me for doing this. I am used to writing it like this in C and Rust.. and I am happy to uphold this style :D 😆

Comment (casperklein) [reply target/scripts/helpers/log.sh:45]: > I always thought people hate me for doing this.

I do 😆 It's just that I don't like wasting a line for no benefits. The some goes for if statements, I prefer for simplicity:

```bash
if true; then
    foo
fi
```

So if there are plans to change that, I am all for it 👍 However I understand that I am not the only maintainer, so my personal style wishes may not always meet those of the others. Most important for me is consistency across our code base, even if I don't like the decision that was chosen 😆

File: target/scripts/helpers/log.sh
Comment (casperklein) [reply target/scripts/helpers/log.sh:121]: `grep "^LOG_LEVEL=" /etc/dms-settings | cut -d "'" -f 2` should also work

File: Dockerfile
Comment (casperklein) [reply Dockerfile:42]: It should be as simple as this:

`COPY --link --from=clamav/clamav:latest /var/lib/clamav /var/lib/clamav`

File: target/scripts/helpers/log.sh
Comment (georglauterbach) [reply target/scripts/helpers/log.sh:45]: @casperklein I have no problem with a PR that re-styled _the whole_ code base at once 😆 But I guess this also has not benefit. I would just sick to the style I imposed upon all of you 😆 😆

Comment (polarathene) [reply target/scripts/helpers/log.sh:45]: > Agreed. I always thought people hate me for doing this.

Oh, totally not a fan, but since you've already converted the code-base I just go with it. Not a big deal.

> Most important for me is consistency across our code base, even if I don't like the decision that was chosen :laughing: 

This.

---

If someone wants to do that mundane task of switching a style through all the files, go for it... Just wait until after a release. I especially don't want to have to rebase my WIP PR against it :sweat_smile:

File: target/scripts/helpers/log.sh
Comment (georglauterbach) [reply target/scripts/helpers/log.sh:121]: > `grep "^LOG_LEVEL=" /etc/dms-settings | cut -d "'" -f 2` should also work

I implemented this in #2754 :)

**Pull request closed without merging**
